### ISSUE TITLE
Scheduler needs to deal with pods without resource limits

### ISSUE THREAD

**Issue by bprashanth** [GoogleCloudPlatform/kubernetes#10242]
Setup a 3 node cluster, it'll look like:
```
NAME                                               READY     REASON    RESTARTS   AGE
elasticsearch-logging-v1-trk5e                     1/1       Running   0          2m
elasticsearch-logging-v1-u0pcb                     1/1       Running   0          2m
fluentd-elasticsearch-e2e-test-beeps-minion-12g9   1/1       Running   0          1m
fluentd-elasticsearch-e2e-test-beeps-minion-5bqu   1/1       Running   0          1m
fluentd-elasticsearch-e2e-test-beeps-minion-jir8   1/1       Running   0          1m
kibana-logging-v1-fgfdf                            1/1       Running   0          2m
kube-dns-v3-n17tg                                  3/3       Running   0          2m
monitoring-heapster-v4-plzg4                       1/1       Running   0          2m
monitoring-influx-grafana-v1-hpic0                 2/2       Running   0          2m
```
3 nodes, 9 containers, in theory we should end up with 3 per node but we rarely do because the node scores are affected by the rate of creation of the fluentd pods. 

Create an rc with 100 pods in a custom namespace and you'll end up with all 100 on the same node. This is because the occupancy score of the one node never catches up to the other 2, because the cluster addons have container cpu and memory populated via the default limit range while the 100 pod rc does not.

**Comment by davidopp** [comment]
In our team meeting we decided that first we need to clarify the resolution of #9356. If we decide to put system pods in a separate namespace, then we'll have namespaces that don't set a LimitRange inherit the one set by the default namespace. If we aren't going to put system pods in a separate namespace, then we will punt this issue to post-1.0.

**Comment by dchen1107** [comment]
See my comment at https://github.com/GoogleCloudPlatform/kubernetes/issues/9356#issuecomment-112304758 

This is one of reasons I am not supporting putting addons to other namespace initially. But now I am in the process of updating all addons containers with a resource limit (I' ve already done with heapster and influxdb yesterday), I am ok with customized but preserved namespace for addons.

**Comment by davidopp** [comment]
There are many options here. I do agree we should do something about this. It's not possible to document "we do spreading, except when we don't" in a way that users will understand. So I think we should implement something that will give something resembling spreading-like behavior as much as possible.

Possible solutions we've discussed are (1) having namespaces without a LimitRange inherit the one from the default namespace, and (2) having the scheduler pretend there's a 100m CPU limit on pods that don't specify one (which is basically the same as (1) but hidden from the user). The only other is that @bprashanth proposed in #10243 simply spreading 0-limit pods. I am not a big fan of that because it can lead to weird scenarios like in a 2-node cluster where one machine is empty and one is running one pod that takes up the whole machine, half of the 0-limit pods will go to the full machine, which seems even worse than putting them all on the empty machine (which is what we would do currently).

@bgrant0607

**Comment by lavalamp** [comment]
Another option we talked about was having a pods-per-core limit for nodes.

**Comment by davidopp** [comment]
Ah, right. That requires adapting #5547 a bit and writing an e2e test for it (currently we only have a unit test; requires #8669). If we do that I'd be in favor of getting rid of LimitRange in the default namespace to simplify the system -- is there any reason we're doing that other than to limit #pods/node?

@gmarek

**Comment by bprashanth** [comment]
Though the default limits only specify cpu it would be odd to not have a pods per mb (to compare custom limits that do). I'm finding it a little hard to explain to myself what that even means, it'll probably have to be a fraction, and it effectively boils down to a twisted version of mb per pod.

**Comment by bprashanth** [comment]
> is there any reason we're doing that other than to limit #pods/node?

max_pods won't scale nicely with size of the node right?

**Comment by davidopp** [comment]
> max_pods won't scale nicely with size of the node right?

What I meant was "is there any reason we're doing the 0.1 LimitRange other than to enforce a max pods per node." (I agree that our current max_pods doens't scale with the size of the node.)

**Comment by davidopp** [comment]
BTW a drawback of (2) over (1) or (3) is that we're back to having no way to cap the # of pods per node, in the non-default namespace.

**Comment by davidopp** [comment]
(which we want to do to avoid overloading Kubelet)

**Comment by yujuhong** [comment]
> "is there any reason we're doing the 0.1 LimitRange other than to enforce a max pods per node." (I agree that our current max_pods doens't scale with the size of the node.)

Just a few thoughts since you asked :)
- There can be pods with user-given resource limits in the system too, so it might be easier for the scheduler to set the default cpu limit per pod (using limitRange or simply defaulting) and compute the available cpu per node.
- In the long run (post v1.0), kubelet may report back to the master the actual resource usage of the pods and/or available resources. Using cpu limits seems to fit the plan (compared with maintaining a separate max pods per node based on dynamic node size determined by current usage).

**Comment by dchen1107** [comment]
cc/ @bgrant0607

**Comment by davidopp** [comment]
> In the long run (post v1.0), kubelet may report back to the master the actual resource usage of the 
> pods and/or available resources. Using cpu limits seems to fit the plan (compared with maintaining a 
> separate max pods per node based on dynamic node size determined by current usage).

Since #5547 makes pods (count) an explicit resource, I think it would be easy for us to report back # pods in use in NodeStatus when we report back usage of cpu, mem, etc.

**Comment by bgrant0607** [comment]
Going back to the beginning: Who's creating a namespace and then launch a 100-replica RC in it?

Namespaces are a relatively advanced concept. Kick-the-tires users aren't going to do it -- that's what the default namespace is for. If we're doing this in our tests, we should be creating LimitRanges for those namespaces OR specifying resources for our pods. If we're not specifying resources somehow, we deserve what we get. 

Limiting pods per node isn't going to help in the common case where the number of pods is less than the max per node. It's necessary, but not sufficient. We need to make it scale by node size eventually. 

What we should do:
1. Assign some cpu on all the addons. If we're not doing this, we're not setting a good example for our users, and we're just asking for trouble. 
2. We could use number of pods as an additional priority function, weighted appropriately. Reporting back the number of pods from Kubelet, as well as total allocated resources, would be useful for kubectl get nodes, also. Eventually we want to use some derived function of resource usage, however, at least for placement of best-effort pods.
3. We could spread by labels (proxy for RCs) and namespace, in addition to services, as discussed in #4971.
4. We should create a resource-management tutorial, which explains things like how to debug OOMs, how to figure out resource usage with heapster, how to set and rollout resource limits, and how to use LimitRange. @vishh 
5. We could create a namespace generator in kubectl that can create, e.g., a LimitRange along with the namespace. @derekwaynecarr 
6. At some point, we should switch the scheduler and kubelet admission control to use request rather than limit. If limit is specified but request is not, as is the case today, we should default request from limit, so this is a backward-compatible change. Then, we should change the LimitRange to set request instead of limit, we should add a default value for memory, and we should update the addons correspondingly.
7. Eventually we should do vertical auto-sizing, both initial and reactive.

**Comment by bgrant0607** [comment]
8. At some point, we should also implement QoS buckets #147 to protect pods with limits/requests from pods without them.

**Comment by davidopp** [comment]
> Who's creating a namespace and then launch a 100-replica RC in it?

That's an extreme example. We're mostly trying to avoid user confusion, for example one node with 2.4 limit running on it and one node with 2.5 limit running on it, and now every new pod they submit (assuming a new namespace) will go onto the same node, and they will become confused. It's not so much an issue of "If we're not specifying resources somehow, we deserve what we get." but if the user doesn't specify resources, is it OK to give semi-mysterious scheduling behavior.

> 1. Assign some cpu on all the addons. If we're not doing this, we're not setting a good example for 
> our users, and we're just asking for trouble. 

Dawn has either already done this or is in the process of doing it.

>  2. We could use number of pods as an additional priority function, weighted appropriately. 

Agreed. This is related to what @bprashanth was doing in #10243 but I think we need something a bit more sophisticated because the approach he used completely ignores the limits of the stuff already on the node (when you're scheduling 0-limit pods).

> Eventually we want to use some derived function of resource usage, however, at least for placement 
> of best-effort pods.

Agreed. We're just trying to figure out what to do until we have that.

> 3. We could spread by labels (proxy for RCs) and namespace, in addition to services, as discussed 
> in #4971.

Agreed, but I think that doesn't really help with this issue.

> 4. We should create a resource-management tutorial, which explains things like how to debug
> OOMs, how to figure out resource usage with heapster, how to set and rollout resource limits, and 
> how to use LimitRange. @vishh 

Agreed, but I think that doesn't really help with this issue.

> 5. We could create a namespace generator in kubectl that can create, e.g., a LimitRange along with 
> the namespace. @derekwaynecarr 

That's an interesting idea, but moving the policy from the server to the client (i.e. the server-based LimitRange inheritance idea vs. the approach you described) means it can't be universally enforced.

6. and 7. are very true, but post-1.0.

So it sounds like your suggestion is to refine #10243 to do something a bit more balanced under different scenarios. We'll have to think about what that algorithm would look like. The only downside is that by doing that instead of LimitRange (or starting to use the Kubelet-based max pods flag), we open the Kubelet up to getting overloaded (since we won't be enforcing any kind of max pods per node in , non-default namespace), which I thought we were trying to avoid. Maybe @dchen1107 can comment on how significant that danger is.

**Comment by davidopp** [comment]
I have no idea why Markdown turned all of my numbers into 1's, but hopefully it's still possible to understand.

**Comment by bprashanth** [comment]
> Namespaces are a relatively advanced concept. Kick-the-tires users aren't going to do it -- that's what the default namespace is for. If we're doing this in our tests, we should be creating LimitRanges for those namespaces OR specifying resources for our pods. If we're not specifying resources somehow, we deserve what we get.

This only makes sense to me now that we're moving the cluster addons out of default. So we're saying: here's a bb gun, you'll shoot your eye out. 

Ideally I would come up with a way to:
1. Surface the user error 
2. Never have it occur (kubectl generator)
3. Handle it when it happens (scheduler changes)

You're only saying forget about 3. How do we do 1, and are we doing 2 for1.0?

> Agreed. This is related to what @bprashanth was doing in #10243 but I think we need something a bit more sophisticated because the approach he used completely ignores the limits of the stuff already on the node (when you're scheduling 0-limit pods).

I though about my pr a little. 

IMO we should never hit the node specified max pods, it should be last line of defense (basically the kubelet saying -- if you totally mess up, know that I can't handle more than this). We should be able to achieve the right spread via least requested and balanced resource. I fear that adding a simpler priority function will mess those 2 up.

If we don't know anything about a pod's resource usage I think it's Ok to shove it somewhere, wait for things to go bad, and move it. This is essentially what'll happen if you end up overcommiting the node with a bogus limit on pods that have none. To that end it doesn't really matter where we apply the *initial* bogus limit, since the node can always update the resource usage of a pod and the scheduler will make later decisions based on this. But adding a count based priority function makes this harder. You will have to weigh the counter higher to have it out score the other 2 in this specific scenario (unless the weights are dynamic).

**Comment by lavalamp** [comment]
> 3. We could spread by labels (proxy for RCs) and namespace, in addition to services, as discussed in #4971.

Hard to see that working (deterministically) if the resource checking takes precedence. Oh, I see @davidopp beat me to it. Yeah, it's a reasonable thing to do, but it doesn't help this problem.

**Comment by gmarek** [comment]
I see that I had missed quite a lot by going to conferences, as it took me some time to understand what's the issue.

IIUC we have a problem because current scheduling algorithm can assign "all new pods" to the single machine if it turns out that it's the one with slightly more free resources than the rest. I'm not quite sure what's the role of the namespace in this picture. Is it only LimitRange related, i.e. sets default limit on resources for Pods?

My 2c is that 0 resources is not very different from 1m resources, so I don't think that requiring "non-zero" requests will solve anything (except "forgot to declare" cases, but that won't be all of them).

Either way I think that the best way to go, at least for now, is to go with @bgrant0607 point (2) / refined @bprashanth PR #10243. E.g. we can lower the score of nodes which are running multiple pods with the same set of labels (kind of @bgrant0607 point (3), but on the scoring, not spreading side of the algorithm). Or is there a reason why we don't want to influence score in this way?

As for why @daviopp brought me here: we can easily change `max_pods` to `max_pods_per_sth` to make it scale with something, e.g. number of cores on the machine. IIRC LimitRange goal was to, in a sense, limit the number of pods/resources used per node per user (namespace). I agree with @davidopp that for default namespace is pretty much the same, as limiting the number of pods per node, so there's no point of having LimitRange for it (except consistency).

**Comment by bgrant0607** [comment]
Part of the problem is that it doesn't make sense for LeastRequested to be the highest-weighted priority for pods that don't request any resources.

**Comment by bgrant0607** [comment]
@gmarek CalculateAntiAffinityPriority is a priority function, so I'm not sure what you mean by "on the scoring, not spreading side".

**Comment by bgrant0607** [comment]
Max. pods should scale with number of nodes, but is irrelevant to this problem. As @bprashanth wrote, we shouldn't be hitting max. pods.

**Comment by bgrant0607** [comment]
I'd frame the problem this way:

We don't want a single node to become a magnet for all pods that specify no require requirements.

Therefore, other criteria must take precedence for such pods than LeastRequested, since the placement of such pods doesn't change future LeastRequested decisions.

**Comment by bgrant0607** [comment]
@davidopp This kind of problem is one reason why vertical auto-sizing is important.

**Comment by bgrant0607** [comment]
cc @justinsb

**Comment by bgrant0607** [comment]
@gmarek No, LimitRange was not just intended to create another max per node. It made the LeastRequested priority function do something, so we'd get scheduling decisions that were better than just random selection. For instance, if all pods had just the default value, they'd be spread evenly. If there were some large-resource containers, that would cause fewer pods to be scheduled on the same node.

**Comment by bgrant0607** [comment]
Previous discussion was in #4910

**Comment by davidopp** [comment]
> This kind of problem is one reason why vertical auto-sizing is important.

I completely agree. The question is what to do until we have that. In a sense, applying a fixed resource limit to every pod that requests zero (i.e. what we're doing with LimitRange in default namespace currently, and arguably should do in all namespaces by default) is a degenerate case of vertical auto-sizing.

**Comment by dchen1107** [comment]
+1 on what @davidopp said above.

**Comment by davidopp** [comment]
I think everyone here is agreeing that the right thing to do (for 1.0) is to enhance @bprashanth 's PR so that it behaves slightly better. 

I agree that whether to add default LimitRange to non-default namespace is somewhat orthogonal because giving 0-limit pods a 0.1 CPU limit doesn't help much especially on beefier nodes (e.g. 2-node cluster with 8-core nodes, one node has a 4-CPU-limit pod running on it, even with 0.1 CPU LimitRange the current scheduler will put 39 0-limit pods on the empty node before it goes back to the one with the 4-CPU-limit pod.) If we can do the spreading algorithm right, then the LimitRange is less of an issue.

I can try to think of what a good priority/scoring algorithm would be for 0-limit pods so it doesn't have the "magnet" problem @bgrant0607 mentioned nor the "totally oblivious to what's already running" that the current PR does.

**Comment by yujuhong** [comment]
This is not directly related to scheduling, but cpu limit would also affect the relative cpu weight when running the docker container. If we don't have a default cpu limit for every pod, each pod will by default have 1024 cpu shares per container. Any pod/container with cpu limit set to be less than 1 core may actually have less cpu time than the defaulted pods on the node.

Assume there are three containers on the node: 
c0: 0.5 cpu limit, c1: no cpu limit, c2: no cpu limit

The cpu shares will be: 
c0: 512 shares, c1: 1024 shares, c2: 1024 shares 

The % of cpu shares (excluding kubelet and the daemons) would become:
c0: 20%, c1: 40%, c2: 40%

It seems unfair to punish the pod/container which has an explicitly set cpu limit in this case.

**Comment by bgrant0607** [comment]
@yujuhong What is setting the default to 1024 shares? A container with no resource request should get a negligible number of shares.

**Comment by yujuhong** [comment]
@bgrant0607, unless I read it wrong, docker sets the default cpu share for every container to 1024 (https://docs.docker.com/reference/run/). This corresponds to the --cpu-share=0 default.

**Comment by davidopp** [comment]
I discussed a little with @bprashanth and he clarified what he thinks @bgrant0607 is proposing. I had a separate idea, so let me present them both here and see which people prefer. I do think we need to do something for 1.0, and time is running out.

@bprashanth's interpretation of @bgrant0607's idea was
* Create a new priority function, call it something like globalSpread (to differentiate it from serviceSpread, which spreads pods belonging to the same service). It does what @bprashanth 's PR currently does for 0-limit pods.
* Change the weighting used in prioritizeNodes() so that instead of being static, it takes into account the characteristics of the pod being scheduled. In this case, when scheduling a new 0-limit pod, we use a mixture of the existing priority functions *and* globalSpread, whereas when scheduling a new pod with limits, we use just the existing priority functions (globalSpread gets a weight of zero). [Note: We never really considered when a "decision tree" approach to scoring, as in Borg, is better or worse than the "weighted sum" approach to scoring we're using in Kubernetes. Now isn't the time to debate this, but it might be interesting to reconsider later.]

This seems reasonable (I like the idea of splitting this new spreading out into a separate function).

My idea, which I think we can combine with the above one, was the following
* Mostly leave @bprashanth 's PR as-is; that is we're still talking just about using an alternative algorithm when the new pod is 0-limit
* But when adding up the number of pods already on a node, weight them by their limit (expressed as fraction of node capacity) if they have one. So for example, a pod that takes up the whole machine might count as five pods, a pod that takes up 80% of the machine as four pods, a pod that takes up 60% of the machine as three pods, a pod that takes up 40% of the machine as two pods, and a pod that takes up less than 40% of the machine as one pod. These numbers are somewhat arbitrary but they "feel" reasonable -- for example, if we have two nodes, one of which has a total of 20% limit on it (and nothing else) and the other which has a total of 80% limit on it (and nothing else), and we get a stream of 0-limit pods on it, we will schedule three 0-limit pods onto the first machine before we start round-robining between the two machines. We can quibble about whether we should be thinking about # of pods or some percentage that's tied to the machine's total capacity but I think the gist would stay the same.

BTW this is very much like giving a default limit to 0-limit pods but (1) there seemed to be strong resistance to applying default LimitRange to non-default namespace (though I'm not sure why), and (2) the result is a bit more reasonable -- 0.1 is essentially "too small" to be reasonable (I would not consider ten limit-0 pods to be equivalent to one limit-1 pod), and the factors I described above address this by effectively weighting 0-limit pods as requiring more than 0.1.

Now, I think these ideas may actually be compatible, in the sense that what I was describing may just be a fancy way to think about how to set the weighting for 0-limit pods in @bgrant0607 's approach. The only drawback is that the math is a bit complicated to reason about since everything gets remapped into the 0-10 range, then weighted, then added. So I need to think about this some more.

**Comment by bgrant0607** [comment]
I was thinking that in the case of a 0-limit pod, utilization-based priority functions would just return 0, making them no-ops.

We could then implement a separate priority function based on the number of pods, similar to @bprashanth's scoreHostByPodCount. 

That would allow the number of pods to affect the scheduling of pods wit limits set, which is desirable in the absence of both QoS-based isolation on the node and usage-derived scheduling heuristics.

As for creating a LimitRange by default when a new namespace is created, I'm supportive of doing that on the client side in a namespace generator.

**Comment by davidopp** [comment]
> I was thinking that in the case of a 0-limit pod, utilization-based priority functions would just return 0, 
> making them no-ops.

I don't think this is good, because it can lead to bad behaviors like if you have a node that is 90% filled in terms of limits of running pods (and is not running anything else), and a node that is 10% filled in terms of limits of running pods (and is not running anything else), then you will evenly distribute new 0-limit pods between the two whereas it seems clear you should schedule at least the first few 0-limit pods all on the 90% filled machine. When scheduling pods with 0 limit, you really want a mix of utilization and pod count. I suggested a simple (albeit not super-elegant) way to do that.

> That would allow the number of pods to affect the scheduling of pods wit limits set, which is desirable 
> in the absence of both QoS-based isolation on the node and usage-derived scheduling heuristics.

Yes, this is a good point which I hadn't considered. (I had only been thinking about scheduling 0-limit pods, not scheduling with-limit pots).

It does seem that a separate priority function is the way to go, but devising a good weighting policy may be tricky. 

BTW I don't think LimitRange in a client is as useful since people may not use that client. If we're going to rely on LimitRange for spreading we really need to default it on the server IMO.

**Comment by bgrant0607** [comment]
True, 0-limit pods should take occupancy by limit into account to some degree, just as pods with limits should take occupancy by pods into account by some degree.

I don't think we should entirely rely on LimitRange. IMO, creating namespaces is a more advanced use case, we should teach such users how to use LimitRange, and we should make it easy to create. While there will be other clients, and I agree with adding functionality server-side in general, the vast majority of such users would use kubectl, so it's reasonable to put the functionality there until we have server-side config generation (#5280). A kubectl namespace generator could generate LimitRange #10495.

**Comment by gmarek** [comment]
I want to chime back to what @yujuhong wrote. IIUC Docker spreads resources that are left after static (i.e. concrete core ids) allocation between containers assuming some weights. This means that given two Nodes, one with one 3-container pod, and one with one 1-container pod, and trying to schedule one 1-container pod it's better to schedule it on the latter (it'll get 50% of resources, instead of 25% when scheduled on the first - assuming everything runs on 1024 shares). So I'd add to @davidopp idea that we should count containers, not pods, if we want to go that way.

Another idea is just trying to maximize amount of resources 0-limit pods would get? This is `1024/sum of shares of all running containers`, somehow scaled to 0-10 range (e.g. "times 10").

Above are just free standing ideas, as I think that putting real lower limit (i.e. LimitRange) on resource request should be the way to approach this problem. Note that there's a big difference between giving some limit to 0-limit pods only for counting priorities, and really giving and accounting them. Currently only way in which we guarantee that CPU is not "oversubscribed" is by strictly checking sum of resources claimed from a given node, which is done on the predicate checking level. As @yujuhong wrote if we allow 0-limit pod on a machine which has some "normal" pods running, we're reducing CPU shares for all "normal" ones by (by half if the machine was full), as the sum of allocated shares per CPU may be way greater than 1024.

**Comment by bgrant0607** [comment]
Forked #10543 for how to set shares when no cpu is specified.

I can't see how "Docker" can do spreading, since it doesn't update cgroups. Cpu shares are just that -- shares.

Counting containers rather than pods -- makes sense.

What we're trying to do is spread pods by resource usage, but without measuring resource usage. :-(  We could model it that way. I'd use some fraction of specified limits (less than 100%) in that case. Then we would need a separate prioritization function. We already have the fit check based solely on limit.

**Comment by bprashanth** [comment]
> It does seem that a separate priority function is the way to go, but devising a good weighting policy may be tricky.

Do we need to do a smart weight policy pre 1.0? If we change the hardcoded weights to a weight function we could eventually expose this through the scheduler api. For now it could be dumb.

**Comment by davidopp** [comment]
I created a spreadsheet to explore this issue. The hypothesis is that we can create a new spreading function (call it DumbSpreadPriority, to distinguish it from ServiceSpreadPriority) and give it an appropriate weight, so that pods will be spread (by number of pods) and will be biased towards machines with low usage after the new pod is placed (as computed by limits).

I used @bprashanth 's scoreHostByPodCount() function as the way to compute this new spreading priority. It produces a priority (range 0-10, as usual) from two inputs: the number of pods on the machine after the new pod is scheduled (which is of course 1+ the number before it is scheduled) and some denominator. In @bprashanth 's PR the denominator is the pod capacity reported in NodeStatus, but I think we want to choose our own value (that will be smaller, but also fixed) in order for this to work well. 

I computed a table whose rows are % of machine full after scheduling and columns are # of pods on the machine after scheduling. This is a simplified model where the priority is computed as `output_of_LeastRequestedPriority +  N * output_of_DumbSpreadPriority` where N is the weight for DumbSpreadPriority (assume the factor for LeastRequestedPriority is 1). Thus there are two parameters for the table: N and the denominator I mentioned in the previous paragraph.

With a denominator of 10 and N=1, this essentially counts each pod as using 10% of the machine. For example, all of the following {% full, # pods} pairs have a weight of 9: {100%, 1} {90%, 2} {80%, 3} {70%, 4} {60%, 5} {50%, 6} {40%, 7} {30%, 8} {20%, 9} {10%, 10} {0%, 11}. This means that we wold equally prefer a machine that is 70% full (by limit) and has 4 pods on it, as a machine that is 30% full (by limit) and has 8 pods on it. This feels vaguely in the right ballpark to me, but it is easy to customize by changing the denominator and N.

The catch with this is that I said it is a simplified model, because we're only considering DumbSpreadPriority and LeastRequestedPriority. By default we currently add together three functions: LeastRequestedPriority, BalancedResourceAllocation, and ServiceSpreadPriority. I am not sure how to model these. The only thing I thought of was to have each cell of my table be <min, max> instead of a value -- computed as <x, x+20> where x is the value currently in the cell. (That is to say, the range of possible values when you take all four algorithms into account is somewhere between the value I currently compute and that value + 20, since each of the other two algorithms contributes a value between 0 and 10.) I produced such a table (well, just a table for x+20, since obviously the table for x is the one I already had) but I wasn't sure what to make of it. You don't need a table to know that the other two algorithms essentially introduce a bunch of randomness into the analysis, and you just need to pick some weighting factors based on how controlling you want each one to be. My random gut opinion is that we should give DumbSpreadPriority a weight of 2 and each of the other ones a weight of 1, to compensate for the fact that there are two limit-based contributors to the total priority.

In general, I think it would be easier for us to reason about priority/scoring if combining the factors were implement as a decision tree as in Borg, rather than just adding up a bunch of weighted numbers. In general, with our current approach in Kubernetes, you run into the complexity described above, plus in principle it's even worse because in theory you can't even know what other policies will be active in the system (and what their weights will be) when the new policy runs. (I'm completely ignoring that last bit, and assuming we know these are the only four policies that will be running, and that we can control their weights.)

**Comment by davidopp** [comment]
(Oh, and to be clear -- I wasn't using "smart weights." I was just adding a new algorithm to the mix and using exactly the same framework we currently have.)

**Comment by gmarek** [comment]
My feeling about complexity of analysis is similar to @davidopp, so I guess we should create a post-v1 issue for rethinking scheduling algorithm (and possibly rename scoring functions: LeastRequestedPriority confuses me all the time). In example currently I've no idea what is the definition of a thing we're trying to optimize for, hence is hard to reason what solution is 'good' for 0-limit case.

Before v1:
* We can extend service spreading to work on replication controllers if Pods are not under the service when being scheduled (e.g. service was not yet created, as in some of our examples such as `cluster-dns`). This does not solve this issue fully, but helps a bit (huge RC with 0-limit pods, but not a number of small RCs). I believe we should do this either way (especially that it's a small change, and makes correct use of k8s simpler). @mwielgus 
* IIUC @bprashanth idea is to schedule on the node with minimal value of resource-usage + X * number of running pods, where each pod is equivalent to some fixed amount of resource usage (plus some noise). 
* IIUC @davidopp idea is to substitute X * number of running pods with a sum of declared pod limits (CPU? CPU and Memory?) times some constant.

What we're looking for is some proxy for Node "load", as we don't have this data published yet. I'm not really sure that (resource-usage + number of running pods) is the best we can do, but it's reasonable. Other idea is to use assigned CPU shares as this proxy, and try to schedule Pod on Node where it'll get most "bonus CPU" coming from amount of CPU shares that will be still available after scheduling. This is an extension of @davidopp idea, by taking into account limit of a Pod being scheduled and computing it's happiness factor.

Happiness factor is how many times more CPU Pod would get if put on this Node.
In practice it'd eventually translate to (left_capacity/(1-left_capacity)), where left_capacity is a % of capacity left on the Node after the Pod would be scheduled there. Because this function can return any positive value we'd need to monotonically map it to 0-10 interval in some reasonable way.

**Comment by dchen1107** [comment]
>We can extend service spreading to work on replication controllers if Pods are not under the service when being scheduled (e.g. service was not yet created, as in some of our examples such as cluster-dns). This does not solve this issue fully, but helps a bit (huge RC with 0-limit pods, but not a number of small RCs). I believe we should do this either way (especially that it's a small change, and makes correct use of k8s simpler). 

I like this idea, and believe we need this spreading algorithms in a long run anyway. 

>IIUC Prashanth B idea is to schedule on the node with minimal value of resource-usage + X * number of running pods, where each pod is equivalent to some fixed amount of resource usage (plus some noise).

Kubelet can report resource-usage at all 3 levels: node, pod and containers, including system daemons. The problem is that there is no API available to retrieve such information after v1beta3 refactory. All those usage data are collected by heapster directly, and there is no pipeline to feed such data to API server and scheduler today. ppl talked about stats collector for a while, but I don't think it is ready yet.

>IIUC David Oppenheimer idea is to substitute X * number of running pods with a sum of declared pod limits (CPU? CPU and Memory?) times some constant.
What we're looking for is some proxy for Node "load", as we don't have this data published yet. I'm not really sure that (resource-usage + number of running pods) is the best we can do, but it's reasonable. Other idea is to use assigned CPU shares as this proxy, and try to schedule Pod on Node where it'll get most "bonus CPU" coming from amount of CPU shares that will be still available after scheduling. This is an extension of David Oppenheimer idea, by taking into account limit of a Pod being scheduled and computing it's happiness factor.

If required, kubelet could reports the allocated CPU shares to describe Node "load'.

**Comment by gmarek** [comment]
I have a PR for RC spreading ready - I'll test it tomorrow and send it for review before Monday.

**Comment by davidopp** [comment]
Sorry my previous message was so long. I'm suggesting something very simple for 1.0:

1) Take @bprashanth 's PR and split the algorithm into a separate priority function (but use 10 as the denominator instead of max pods reported by NodeStatus)
2) Weight that priority function with weight 2 (all the others are currently weighted as 1)

And that's it. I think it's good enough.

**Comment by gmarek** [comment]
SG for v1, but we should revisit afterwards. In my message I was referring to your message one before the last long one.

What's your take on extending Service spreading to RC? The problem is that some of examples do not care about the order of RC/Service creation, even though it's crucial for the current implementation of Scheduler.

**Comment by davidopp** [comment]
Yes, these problems should go away after v1. Having node-level resource usage information available in the scheduler will help a lot, as will having initial limit guess from auto-scaler.

As for your idea of expending Service spreading to RC -- I guess it depends on the performance and complexity impact. Feel free to put together a PR to do that and we can see how complicated it is.

**Comment by erictune** [comment]
Trying to recap Brian's list of suggestions (which David agreed with), and their current status and if they have issues:

1. *Assign some cpu on all the addons.*    **Done** by #10653
1. this one had two parts:
  1. *use number of pods as ... priority.*  **Not needed** due to #10661.
  1. *use resource usage ... for placement.*  **Not needed** IMHO because it will be done by auto-sizer.
1. *spread by labels (proxy for RCs) and namespace* : **Covered** by #4971
1. *create a resource-management tutorial*:  **Needed, and need to file an issue for it**.
1. *namespace generator [that makes a] LimitRange*: **Not needed** IMO now that we have #10661 
1. *switch the scheduler ... to use request rather than limit*: **Not needed** IMO: sounds more like a primitive vertical auto-sizer, which deserves its own resource, not a LimitRange.
1. *do vertical auto-sizing, both initial and reactive*: **Covered** by #1629. 

@davidopp @bgrant0607 take a look at the above summary and let me know if you agree, and I will file issues where needed for remaining items.

**Comment by bgrant0607** [comment]
(4) is the only thing critical for 1.0. Yes, please file an issue.

(5) is still desirable as a user convenience, independently of this issue, but isn't critical. Covered by #10495

(6) is still desirable independently of this issue, but isn't happening for 1.0. Covered by #168, #147, #7355.

We don't really have an open issue specifically for vertical auto-sizing. We should.

**Comment by davidopp** [comment]
I agree with the list, with the caveat that some of the "not needed" is more like "maybe after 1.0, but not relevant to this issue."

#10667 is related to item 3

**Comment by dchen1107** [comment]
I just filed #10779 for item 4)

**Comment by erictune** [comment]
Filed #10781 for the tutorial.  Filed #10782 for Vertical auto-sizer.

**Comment by gmarek** [comment]
Some thoughts about @davidopp solution to this issue #10661. If we expect to have a lot of 0-limit pods in cluster, current solution may be problematic. What it does in practice in cluster with a lot of running 0-limit pods is to pretty much zero the resource-related score of all machines. This means that if some non-0-limit pod will come it will get assigned random node with enough resources (not the best-fitting one).

**Comment by erictune** [comment]
What does "best-fitting" mean in a cluster with many 0-limit pods?  0-limit
doesn't mean they need no resources, but that their resource usage is
uncertain.  When free resources are highly uncertain, then the "best
fitting" machine is not very different from a random machine.

**Comment by erictune** [comment]
In summary, I think David's solution is good, and that we should not try to
be more clever in the face of uncertainty.

**Comment by gmarek** [comment]
Best fitting for non-0-limit pods. Davids PR assumes 0.1 CPU for each 0-limit pod for scheduling any pod, i.e. when we're looking for pod requiring 0.5 CPU, Nodes with 10 0-limit pods is equivalent to pod with 1 0.5 limit pod and 4 0-limit pods.

I'm not saying we should change it before v1. I'm saying that it's probably not desired behavior.

**Comment by davidopp** [comment]
It's really hard to do anything sensible for 0-limit pods. They could just as easily use 0.01 core as 16 cores on a 16-core machine. I considered several other ideas (including ideas suggested by you, Prashanth, Eric, Brian, etc.) and these alternatives all seemed at least slightly worse. (I should say I think Prashanth was pushing for this approach all along, though the thing he implemented was different.) Now, you could argue that this is the right approach but the limit we give should be higher (e.g. 0.5 core and 100 MB of memory, or whatever) but I chose the numbers based on (1) the fact that LimitRange in default namespace is giving 0.1 core, and (2) the fact that the default limits for the cluster addon pods are close-ish to the values I chose here.

I think the correct post-1.0 solution is to use resource usage in scheduling, plus auto-sizing of, etc.

**Comment by davidopp** [comment]
Oh I forgot, one other consideration I gave was that f1-micro only has 0.6GB of RAM and although we don't use f1-micro by default, some people might use it for ondes, so I didn't want to use default RAM that was too large a fraction of an f1-micro. Otherwise I might have set the default RAM a little higher.

**Comment by gmarek** [comment]
I can't find PR, what happened with the idea of changing default 1024 CPU shares that 0-limit pods get?

**Comment by davidopp** [comment]
It's #10655
